### PR TITLE
Reduce allocations in `BinaryReader`.

### PR DESCRIPTION
**Pull request by teo-tsirpanis**
This PR optimizes the `BinaryReader` class to allocate its fields only if we call its text-related methods. Additionally `_buffer` and `_charBytes` fields were removed and replaced with stack allocations or array pool rents.

### CODE REVIEW

File: src/libraries/System.Private.CoreLib/src/System/IO/BinaryReader.cs
Comment (reflectronic) [inline src/libraries/System.Private.CoreLib/src/System/IO/BinaryReader.cs:475]:
```suggestion
                result = result[..numRead];
```

File: src/libraries/System.Private.CoreLib/src/System/IO/BinaryReader.cs
Comment (stephentoub) [inline src/libraries/System.Private.CoreLib/src/System/IO/BinaryReader.cs:67]: If I'm reading the code correctly, we'll now never use the _buffer outside of FillBuffer and never consume the data read into the buffer.  Can't we just delete _bufferSize and _buffer entirely, and just use ArrayPool in FillBuffer if/when we need a temporary array?

File: src/libraries/System.Private.CoreLib/src/System/IO/BinaryReader.cs
Comment (teo-tsirpanis) [reply src/libraries/System.Private.CoreLib/src/System/IO/BinaryReader.cs]: Thought it was less efficient but [it wasn't](https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA0AXEBDAzgWwB8ABAJgEYBYAKGIGYACY8pJ0hgYQYG8aH+B/ekxYNgATwwwA2gF0GAJRjYAJgCFJMXAAoAlgDsMDSAFdDaBgaOnDpAJR9BA3tSdvLAMwbabRgLx+DAAMDq7ugi7h4cQA7AwAglBQ2OIAdACi+AAOGOIAPBJSAHzadgDcjlEMAL40lVGFMvKwuCYANv4M+jAA7mKa0r6yFWFVul4+EGYY7ACEgS3tGKkAMjD6AOYYABahVfyR+4IA9McMACpQuvgM2EkpqRfburgMuNtTbSoM29hZWesGBB9Ax0gB5ABiDAAZAwshBcLhdMA2uIujAjLgMLBsPhcKl6kdFh0GAstEtUvFcABlLLYfSlaSpVK+UiyVLnCCJZLiUojI61aiE9yxBjEjD89yC6KMZisRpyRTKdSaXCkPSGYxTcyWTWsvbuQ5jCa+UmBELCtxGo5MOLch6ZHL5RolcqWpyC92CBXNckkwLdPoKoaS43eVkMeZiv3LNabHYGqrWo6nC5XG53HmPc7PV7vT7fX7/QHA0GQmFwhFIlFo7qY7HKPEE0Y28Wk6OtDpMlnambDL0CT0t6JxcWhj00apAA==), thanks!

File: src/libraries/System.Private.CoreLib/src/System/IO/BinaryReader.cs
Comment (teo-tsirpanis) [reply src/libraries/System.Private.CoreLib/src/System/IO/BinaryReader.cs]: Didn't think of that, thanks. `FillBuffer` has also been refactored to avoid it if length is zero or one.

File: src/libraries/System.Private.CoreLib/src/System/IO/BinaryReader.cs
Comment (teo-tsirpanis) [inline src/libraries/System.Private.CoreLib/src/System/IO/BinaryReader.cs:290]: The use of a stateful class like `Decoder` in this method does not seem right to me. Isn't each invocation of `ReadString` standalone?

**Comment by adamsitnik** [comment]
@teo-tsirpanis thank you for your contribution! Could you please provide benchmark results that show the gains from the proposed changes? Thanks!

**Comment by teo-tsirpanis** [comment]
I will, let's address https://github.com/dotnet/runtime/pull/80331#discussion_r1064051030 first.

**Comment by jeffhandley** [comment]
@teo-tsirpanis - It looks like you're blocked and awaiting guidance from @stephentoub or @adamsitnik on the [question you posted here](https://github.com/dotnet/runtime/pull/80331#discussion_r1064051030); is that the accurate status of this PR?

**Comment by teo-tsirpanis** [comment]
Yes. And I can't work on it for the next couple of days.

File: src/libraries/System.Private.CoreLib/src/System/IO/BinaryReader.cs
Comment (adamsitnik) [reply src/libraries/System.Private.CoreLib/src/System/IO/BinaryReader.cs]: From the [docs]( https://learn.microsoft.com/en-us/dotnet/api/system.io.binaryreader.readstring?view=net-7.0):

> Reads a string from the current stream. The string is prefixed with the length, encoded as an integer seven bits at a time.

My understanding is that in this particular case the string must be prefixed with the length, so the job of the method is to read the bytes from the stream (and possibly modify it’s state like Position) and covert them into string.


The method does not read all bytes at a time (`readLength` can be different than `stringLength`) so the encoder should maintain it's state between the loop iterations:

https://github.com/dotnet/runtime/blob/be84df7642f21702b0e7859f0c134e4d6cfc67eb/src/libraries/System.Private.CoreLib/src/System/IO/BinaryReader.cs#L272

https://github.com/dotnet/runtime/blob/be84df7642f21702b0e7859f0c134e4d6cfc67eb/src/libraries/System.Private.CoreLib/src/System/IO/BinaryReader.cs#L289-L291

Now should it reset the state in the final iteration?

Checking for the leftovers and throwing would be considered a breaking change. Just resetting the state would be considered a breaking change as well, because someone in theory could rely on keeping this state.

File: src/libraries/System.Private.CoreLib/src/System/IO/BinaryReader.cs
Comment (adamsitnik) [inline src/libraries/System.Private.CoreLib/src/System/IO/BinaryReader.cs:235]: Please run a benchmark that reads floats from `BinaryReader`.

File: src/libraries/System.Private.CoreLib/src/System/IO/BinaryReader.cs
Comment (adamsitnik) [inline src/libraries/System.Private.CoreLib/src/System/IO/BinaryReader.cs:223]: subjective nit: in theory everyone should know the sizes of primitive types, but on the other hand it looks just 100% self-describing when using `typeof(T)` instead consts.

```suggestion
        public virtual short ReadInt16() => BinaryPrimitives.ReadInt16LittleEndian(InternalRead(stackalloc byte[typeof(short)]));
```

File: src/libraries/System.Private.CoreLib/src/System/IO/BinaryReader.cs
Comment (adamsitnik) [inline src/libraries/System.Private.CoreLib/src/System/IO/BinaryReader.cs:292]: it would be great to benchmark this test case as well

File: src/libraries/System.Private.CoreLib/src/System/IO/BinaryReader.cs
Comment (adamsitnik) [inline src/libraries/System.Private.CoreLib/src/System/IO/BinaryReader.cs:533]: The description of this method says:

> Fills the internal buffer with the specified number of bytes read from the stream.

These changes are breaking the public contract as they don't fill the internal buffer.

Our test coverage for it is far from ideal:

https://github.com/dotnet/runtime/blob/be84df7642f21702b0e7859f0c134e4d6cfc67eb/src/libraries/System.IO/tests/BinaryReader/BinaryReaderTests.cs#L441-L462

I know that we should most likely not expose such method in the first place, as it limits the number of optimizations we can apply, but breaking it is imo not acceptable.

@stephentoub what is your opinion on this matter?

**Review by adamsitnik — CHANGES_REQUESTED** [review]
thank you for your contribution @teo-tsirpanis ! I would be fine with the proposed changed if `FillBuffer` was not part of the public API.

File: src/libraries/System.Private.CoreLib/src/System/IO/BinaryReader.cs
Comment (teo-tsirpanis) [reply src/libraries/System.Private.CoreLib/src/System/IO/BinaryReader.cs:533]: This "internal buffer" that gets filled is otherwise unused by the rest of the code and inaccessible to the user. `FillBuffer` still satisfies its contract by reading from the stream.

Comment (stephentoub) [reply src/libraries/System.Private.CoreLib/src/System/IO/BinaryReader.cs:530]: > what is your opinion on this matter?

Is there any way in which the change is visible via public API?  If the answer is no, I think it's fine.  We should also obsolete FillBuffer separately.

File: src/libraries/System.Private.CoreLib/src/System/IO/BinaryReader.cs
Comment (danmoseley) [inline src/libraries/System.Private.CoreLib/src/System/IO/BinaryReader.cs:528]: in general I believe we don't attempt to return the buffer in exceptional cases -- saving the try/catch. Is this a case where exceptions are sufficiently unexceptional that it's worth having try/catch?

BTW I don't know whether try/catch ends up measurably impacting perf, but it presumably at least increases generated code length.

File: src/libraries/System.Private.CoreLib/src/System/IO/BinaryReader.cs
Comment (teo-tsirpanis) [reply src/libraries/System.Private.CoreLib/src/System/IO/BinaryReader.cs]: > Shouldn't we `Reset` it on each non-text operation?

I mean, if we call `Read()`, it will return the next character, and if we call `ReadUInt32(); Read()` after that, the decoder's state will be out-of-sync from the stream.

**Comment by danmoseley** [comment]
Hi @teo-tsirpanis are you still working on this one? I see there's merge conflicts.

File: src/libraries/System.Private.CoreLib/src/System/IO/BinaryReader.cs
Comment (adamsitnik) [reply src/libraries/System.Private.CoreLib/src/System/IO/BinaryReader.cs]: > I mean, if we call Read(), it will return the next character, and if we call ReadUInt32(); Read() after that, the decoder's state will be out-of-sync from the stream.

Could you please add a unit test that exhibits the described behavior?

File: src/libraries/System.Private.CoreLib/src/System/IO/BinaryReader.cs
Comment (adamsitnik) [reply src/libraries/System.Private.CoreLib/src/System/IO/BinaryReader.cs:530]: > Is there any way in which the change is visible via public API? If the answer is no, I think it's fine.

Thansk, in such case I am fine with it too.

File: src/libraries/System.Private.CoreLib/src/System/IO/BinaryReader.cs
Comment (adamsitnik) [inline src/libraries/System.Private.CoreLib/src/System/IO/BinaryReader.cs:521]: In such case we can simply seek by `numBytes`-bytes?

```suggestion
                    if (_stream.CanSeek)
                    {
                        _stream.Seek(numBytes, SeekOrigin.Current);
                        return;
                    }

                    byte[] buffer = ArrayPool<byte>.Shared.Rent(numBytes);
```

**Comment by teo-tsirpanis** [comment]
Thanks for the review, will take me some time until I can address it.

**Comment by adamsitnik** [comment]
> Thanks for the review, will take me some time until I can address it.

No worries! I'll re-apply `needs-author-action` label so this PR does not pop up in the list of stale PRs. Thank you!

File: src/libraries/System.Private.CoreLib/src/System/IO/BinaryReader.cs
Comment (teo-tsirpanis) [reply src/libraries/System.Private.CoreLib/src/System/IO/BinaryReader.cs]: Nice idea, thanks! Fixed.

File: src/libraries/System.Private.CoreLib/src/System/IO/BinaryReader.cs
Comment (teo-tsirpanis) [reply src/libraries/System.Private.CoreLib/src/System/IO/BinaryReader.cs]: Done in dotnet/performance#3114.

File: src/libraries/System.Private.CoreLib/src/System/IO/BinaryReader.cs
Comment (teo-tsirpanis) [reply src/libraries/System.Private.CoreLib/src/System/IO/BinaryReader.cs:289]: Done in dotnet/performance#3114.

**Comment by adamsitnik** [comment]
@teo-tsirpanis could you please share the benchmark results?

**Comment by teo-tsirpanis** [comment]
I will do it in the weekend. When will the snap for .NET 9 happen?

**Comment by adamsitnik** [comment]
> I will do it in the weekend. When will the snap for .NET 9 happen?

The snap where we could merge all changes already happened on July 17th. Until August 14th we can merge bug fixes, important features and important perf fixes, but we need a direct manager approval. After Aug 14th only impactful issues, with more complex approval process.

**Comment by teo-tsirpanis** [comment]
Here they are:

`runtime-old` is 41b8bc43b675790d68c5d39d4f8dad3f1a33bee0 (the commit from `main` we merged into)
`runtime-new` is 62732011ffb82c46c9a054e408ce2cf0b420c3bd

``` ini

BenchmarkDotNet=v0.13.2.2052-nightly, OS=Windows 10 (10.0.19045.3271)
Intel Core i7-7700HQ CPU 2.80GHz (Kaby Lake), 1 CPU, 8 logical and 4 physical cores
.NET SDK=8.0.100-preview.6.23330.14
  [Host]     : .NET 8.0.0 (8.0.23.32907), X64 RyuJIT AVX2
  Job-WTHNYV : .NET 8.0.0 (42.42.42.42424), X64 RyuJIT AVX2
  Job-BHRFRA : .NET 8.0.0 (42.42.42.42424), X64 RyuJIT AVX2

PowerPlanMode=00000000-0000-0000-0000-000000000000  Arguments=/p:EnableUnsafeBinaryFormatterSerialization=true  IterationTime=250.0000 ms  
MaxIterationCount=20  MinIterationCount=15  WarmupCount=1  

```
|           Method |        Job |                                                   Toolchain |       Mean |      Error |     StdDev |     Median |        Min |        Max | Ratio | RatioSD |   Gen0 | Allocated | Alloc Ratio |
|----------------- |----------- |------------------------------------------------------------ |-----------:|-----------:|-----------:|-----------:|-----------:|-----------:|------:|--------:|-------:|----------:|------------:|
|      DefaultCtor | Job-WTHNYV | \runtime-new\shared\Microsoft.NETCore.App\8.0.0\corerun.exe |  17.602 ns |  1.1102 ns |  1.2785 ns |  17.796 ns |  15.588 ns |  19.620 ns |  0.54 |    0.05 | 0.0178 |      56 B |        0.35 |
|      DefaultCtor | Job-BHRFRA | \runtime-old\shared\Microsoft.NETCore.App\8.0.0\corerun.exe |  32.702 ns |  1.3043 ns |  1.5020 ns |  32.652 ns |  30.358 ns |  35.462 ns |  1.00 |    0.00 | 0.0510 |     160 B |        1.00 |
|                  |            |                                                             |            |            |            |            |            |            |       |         |        |           |             |
|         ReadBool | Job-WTHNYV | \runtime-new\shared\Microsoft.NETCore.App\8.0.0\corerun.exe |   1.346 ns |  0.1610 ns |  0.1790 ns |   1.350 ns |   1.133 ns |   1.777 ns |  0.88 |    0.15 |      - |         - |          NA |
|         ReadBool | Job-BHRFRA | \runtime-old\shared\Microsoft.NETCore.App\8.0.0\corerun.exe |   1.572 ns |  0.1748 ns |  0.1795 ns |   1.528 ns |   1.340 ns |   2.028 ns |  1.00 |    0.00 |      - |         - |          NA |
|                  |            |                                                             |            |            |            |            |            |            |       |         |        |           |             |
|    ReadAsciiChar | Job-WTHNYV | \runtime-new\shared\Microsoft.NETCore.App\8.0.0\corerun.exe |  17.657 ns |  0.6939 ns |  0.7425 ns |  17.350 ns |  16.933 ns |  19.590 ns |  0.86 |    0.06 |      - |         - |          NA |
|    ReadAsciiChar | Job-BHRFRA | \runtime-old\shared\Microsoft.NETCore.App\8.0.0\corerun.exe |  20.638 ns |  0.8774 ns |  0.9010 ns |  20.428 ns |  19.445 ns |  22.915 ns |  1.00 |    0.00 |      - |         - |          NA |
|                  |            |                                                             |            |            |            |            |            |            |       |         |        |           |             |
| ReadNonAsciiChar | Job-WTHNYV | \runtime-new\shared\Microsoft.NETCore.App\8.0.0\corerun.exe | 102.504 ns |  3.9581 ns |  4.5582 ns | 102.609 ns |  93.493 ns | 109.382 ns |  1.03 |    0.04 |      - |         - |          NA |
| ReadNonAsciiChar | Job-BHRFRA | \runtime-old\shared\Microsoft.NETCore.App\8.0.0\corerun.exe |  99.448 ns |  2.4639 ns |  2.5302 ns |  99.793 ns |  94.699 ns | 104.341 ns |  1.00 |    0.00 |      - |         - |          NA |
|                  |            |                                                             |            |            |            |            |            |            |       |         |        |           |             |
|       ReadUInt16 | Job-WTHNYV | \runtime-new\shared\Microsoft.NETCore.App\8.0.0\corerun.exe |   4.289 ns |  0.2605 ns |  0.3000 ns |   4.200 ns |   3.760 ns |   4.873 ns |  0.75 |    0.06 |      - |         - |          NA |
|       ReadUInt16 | Job-BHRFRA | \runtime-old\shared\Microsoft.NETCore.App\8.0.0\corerun.exe |   5.701 ns |  0.1344 ns |  0.1191 ns |   5.707 ns |   5.545 ns |   5.977 ns |  1.00 |    0.00 |      - |         - |          NA |
|                  |            |                                                             |            |            |            |            |            |            |       |         |        |           |             |
|       ReadUInt32 | Job-WTHNYV | \runtime-new\shared\Microsoft.NETCore.App\8.0.0\corerun.exe |   4.004 ns |  0.2161 ns |  0.2489 ns |   3.982 ns |   3.619 ns |   4.481 ns |  0.73 |    0.05 |      - |         - |          NA |
|       ReadUInt32 | Job-BHRFRA | \runtime-old\shared\Microsoft.NETCore.App\8.0.0\corerun.exe |   5.475 ns |  0.1395 ns |  0.1493 ns |   5.469 ns |   5.302 ns |   5.792 ns |  1.00 |    0.00 |      - |         - |          NA |
|                  |            |                                                             |            |            |            |            |            |            |       |         |        |           |             |
|       ReadUInt64 | Job-WTHNYV | \runtime-new\shared\Microsoft.NETCore.App\8.0.0\corerun.exe |   3.978 ns |  0.2170 ns |  0.2499 ns |   3.960 ns |   3.577 ns |   4.521 ns |  0.73 |    0.05 |      - |         - |          NA |
|       ReadUInt64 | Job-BHRFRA | \runtime-old\shared\Microsoft.NETCore.App\8.0.0\corerun.exe |   5.455 ns |  0.1771 ns |  0.1969 ns |   5.505 ns |   5.137 ns |   5.744 ns |  1.00 |    0.00 |      - |         - |          NA |
|                  |            |                                                             |            |            |            |            |            |            |       |         |        |           |             |
|         ReadHalf | Job-WTHNYV | \runtime-new\shared\Microsoft.NETCore.App\8.0.0\corerun.exe |   5.168 ns |  0.5830 ns |  0.6480 ns |   4.943 ns |   4.483 ns |   6.770 ns |  0.83 |    0.11 |      - |         - |          NA |
|         ReadHalf | Job-BHRFRA | \runtime-old\shared\Microsoft.NETCore.App\8.0.0\corerun.exe |   6.138 ns |  0.1526 ns |  0.1428 ns |   6.159 ns |   5.866 ns |   6.360 ns |  1.00 |    0.00 |      - |         - |          NA |
|                  |            |                                                             |            |            |            |            |            |            |       |         |        |           |             |
|       ReadSingle | Job-WTHNYV | \runtime-new\shared\Microsoft.NETCore.App\8.0.0\corerun.exe |   4.972 ns |  0.1933 ns |  0.2226 ns |   4.944 ns |   4.573 ns |   5.332 ns |  0.81 |    0.08 |      - |         - |          NA |
|       ReadSingle | Job-BHRFRA | \runtime-old\shared\Microsoft.NETCore.App\8.0.0\corerun.exe |   6.138 ns |  0.3766 ns |  0.4337 ns |   5.993 ns |   5.514 ns |   7.003 ns |  1.00 |    0.00 |      - |         - |          NA |
|                  |            |                                                             |            |            |            |            |            |            |       |         |        |           |             |
|       ReadDouble | Job-WTHNYV | \runtime-new\shared\Microsoft.NETCore.App\8.0.0\corerun.exe |   4.903 ns |  0.1319 ns |  0.1466 ns |   4.886 ns |   4.689 ns |   5.207 ns |  0.77 |    0.07 |      - |         - |          NA |
|       ReadDouble | Job-BHRFRA | \runtime-old\shared\Microsoft.NETCore.App\8.0.0\corerun.exe |   6.379 ns |  0.5360 ns |  0.5958 ns |   6.387 ns |   5.456 ns |   7.763 ns |  1.00 |    0.00 |      - |         - |          NA |
|                  |            |                                                             |            |            |            |            |            |            |       |         |        |           |             |
|  ReadSmallString | Job-WTHNYV | \runtime-new\shared\Microsoft.NETCore.App\8.0.0\corerun.exe |  42.309 ns |  1.2379 ns |  1.3759 ns |  42.268 ns |  40.002 ns |  44.566 ns |  0.88 |    0.07 | 0.0153 |      48 B |        1.00 |
|  ReadSmallString | Job-BHRFRA | \runtime-old\shared\Microsoft.NETCore.App\8.0.0\corerun.exe |  47.916 ns |  3.2521 ns |  3.7452 ns |  47.150 ns |  43.816 ns |  56.249 ns |  1.00 |    0.00 | 0.0151 |      48 B |        1.00 |
|                  |            |                                                             |            |            |            |            |            |            |       |         |        |           |             |
|  ReadLargeString | Job-WTHNYV | \runtime-new\shared\Microsoft.NETCore.App\8.0.0\corerun.exe | 346.669 ns | 16.3936 ns | 18.8789 ns | 344.723 ns | 315.138 ns | 391.731 ns |  0.90 |    0.09 | 0.8388 |    2632 B |        1.00 |
|  ReadLargeString | Job-BHRFRA | \runtime-old\shared\Microsoft.NETCore.App\8.0.0\corerun.exe | 385.614 ns | 22.4324 ns | 24.9335 ns | 377.237 ns | 354.246 ns | 432.387 ns |  1.00 |    0.00 | 0.8385 |    2632 B |        1.00 |

File: src/libraries/System.Private.CoreLib/src/System/IO/BinaryReader.cs
Comment (teo-tsirpanis) [reply src/libraries/System.Private.CoreLib/src/System/IO/BinaryReader.cs]: I can't find an easy way to test it. Can we skip it for this PR?

File: src/libraries/System.Private.CoreLib/src/System/IO/BinaryReader.cs
Comment (adamsitnik) [inline src/libraries/System.Private.CoreLib/src/System/IO/BinaryReader.cs:500]: Please remove `ArgumentOutOfRange_BinaryReaderFillBuffer` resource from 

https://github.com/dotnet/runtime/blob/e2b5936b758c10fe4f4f287393cb14a288febbd7/src/libraries/System.Private.CoreLib/src/Resources/Strings.resx#L1735-L1737

File: src/libraries/System.Private.CoreLib/src/System/IO/BinaryReader.cs
Comment (adamsitnik) [inline src/libraries/System.Private.CoreLib/src/System/IO/BinaryReader.cs:499]: nit: it seems like this comment is no longer valid (you have found the threshold)

```suggestion
```

**Review by adamsitnik — COMMENTED** [review]
Overall it looks good, I found only one minor thing (unused resource can be removed).

Thank you for your contribution @teo-tsirpanis !

**Comment by teo-tsirpanis** [comment]
Feedback addressed.

**Review by adamsitnik — APPROVED** [review]
LGTM! Now I'll have to discuss with @jeffhandley whether we want to include this in .NET 8 or 9, will keep you updated.

Thank you again @teo-tsirpanis !

**Comment by jeffhandley** [comment]
I'd prefer to let this wait until .NET 9. It's a change in low-level IO dealing with serialization, with a small bit of `unsafe` added in, and it wasn't something we had planned for .NET 8.

**Comment by teo-tsirpanis** [comment]
@adamsitnik can we merge it?

**Comment by EgorBo** [comment]
Improvements on x64 https://github.com/dotnet/perf-autofiling-issues/issues/23341

**Pull request merged**
